File: cpp/Linked_List/0160_Intersection_of_Two_Linked_Lists.cpp
using namespace std;

struct ListNode {
  int val;
  ListNode* next;

  ListNode(int x) : val(x), next(NULL) {}
};
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
  /*
   * 1st solution length difference
   * */
  ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
    int lenA = 0, lenB = 0;
    ListNode* p1 = headA;
    ListNode* p2 = headB;
    while (p1) {
      p1 = p1->next;
      lenA++;
    }
    while (p2) {
      p2 = p2->next;
      lenB++;
    }
    int lenDiff = lenA - lenB;
    if (lenDiff > 0) {
      while (lenDiff > 0) {
        headA = headA->next;
        lenDiff--;
      }
    } else if (lenDiff < 0) {
      while (lenDiff < 0) {
        headB = headB->next;
        lenDiff++;
      }
    }

    while (headA && headB) {
      if (headA->val == headB->val) return p1;
      headA = headA->next;
      headB = headB->next;
    }
    return nullptr;
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> [ListNode]:
        """
        find the length of A and B
        longer List go first by steps(A-B)
        move forward together
        """
        # len_a = len_b = 0
        # p_a, p_b = headA, headB
        # while p_a:
        #     p_a = p_a.next
        #     len_a += 1
        # while p_b:
        #     p_b = p_b.next
        #     len_b += 1
        # cur_a, cur_b = headA, headB
        # if len_b > len_a:
        #     len_a, len_b = len_b, len_a
        #     cur_a, cur_b = cur_b, cur_a
        # step = len_a - len_b
        # while step:
        #     cur_a = cur_a.next
        #     step -= 1
        # while cur_a:
        #     if cur_a == cur_b:
        #         return cur_a
        #     cur_a = cur_a.next
        #     cur_b = cur_b.next

        """
        concatenate A and B
        if there is an intersection, then it's a loop
        if no intersection, then no loop, both list hit None 
        """
        if headA and headB:
            cur_a, cur_b = headA, headB
            while cur_a is not cur_b:
                cur_a = cur_a.next if cur_a else headB
                cur_b = cur_b.next if cur_b else headA
            return cur_a
